gateway-ha/src/main/java/io/trino/gateway/ha/clustermonitor/ClusterStatsMetricsMonitor.java

```diff
@@ -0,0 +1,198 @@
+        for (String metric : metricNames) {
+            uri.addParameter("name[]", metric);
+        }
+
+        Request request = prepareGet()
+                .setUri(uri.build())
+                .addHeader(identityHeader.name, identityHeader.value)
+                .addHeader("Content-Type", "application/openmetrics-text; version=1.0.0; charset=utf-8")
+                .build();
+        try {
+            return client.execute(request, metricsResponseHandler);
+        }
+        catch (UnexpectedResponseException e) {
+            if (shouldRetry(e.getStatusCode())) {
+                if (retriesRemaining > 0) {
+                    log.warn("Retrying health check on error: %s, ", e.toString());
+                    return getMetrics(baseUrl, retriesRemaining - 1);
+                }
+                log.error("Encountered error %s, no retries remaining", e.toString());
+            }
+            log.error(e, "Health check failed with non-retryable response. %s\n%s", e.getMessage(), e.toString());
+        }
+        catch (Exception e) {
+            log.error(e, "Exception checking %s for health", request.getUri());
+        }
+        return ImmutableMap.of();
+    }
+
+    private static class MetricsResponseHandler
+            implements ResponseHandler<Map<String, String>, RuntimeException>
+    {
+        private final ImmutableSet<String> requiredKeys;
+
+        public MetricsResponseHandler(ImmutableSet<String> requiredKeys)
+        {
+            this.requiredKeys = requireNonNull(requiredKeys);
+        }
+
+        @Override
+        public Map<String, String> handleException(Request request, Exception exception)
+                throws RuntimeException
+        {
+            throw propagate(request, exception);
+        }
+
+        @Override
+        public Map<String, String> handle(Request request, Response response)
+                throws RuntimeException
+        {
+            try {
+                String responseBody = new String(response.getInputStream().readAllBytes(), UTF_8);
+                Map<String, String> metrics = Arrays.stream(responseBody.split("\n"))
+                        .filter(s -> !s.startsWith("#"))
+                        .collect(toImmutableMap(s -> s.split(" ")[0], s -> s.split(" ")[1]));
+                if (!metrics.keySet().containsAll(requiredKeys)) {
+                    throw new UnexpectedResponseException(
+                            format("Request is missing required keys: \n%s\nin response: '%s'", String.join("\n", requiredKeys), responseBody),
+                            request,
+                            response);
+                }
+                return metrics;
+            }
+            catch (IOException e) {
+                throw new UnexpectedResponseException(request, response);
+            }
+        }
+    }
+
+    private record Header(String name, String value) {}
```

Comment:
Can this be moved be at the top for better understanding of the code?

Reply:
agree with @ebyhr here, inner classes should be at the bottom

Reply:
agreed with @ebyhr here, since this is a subclass